digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/AwsService.scala

```diff
@@ -0,0 +1,299 @@
+              speedLimitService.update(value.id, Seq(newSpeedLimitAsset), AwsUser)
+            case None =>
+              val newSpeedLimitAsset = NewLinearAsset(link.linkId, 0, link.length, NumericValue(speedLimit.toInt), properties("sideCode").asInstanceOf[BigInt].intValue(), vvhClient.roadLinkData.createVVHTimeStamp(), None)
+              speedLimitService.createMultiple(Seq(newSpeedLimitAsset), speedLimit.toInt, AwsUser, vvhClient.roadLinkData.createVVHTimeStamp(), (_, _) => Unit)
+          }
+      }
+
+      pavementClass match {
+        case "" =>
+        case _ =>
+          val pavementClassValue = DynamicValue(DynamicAssetValue(Seq(DynamicProperty("paallysteluokka", "single_choice", required = false, Seq(DynamicPropertyValue(pavementClass.toInt))))))
+
+          val duplicate = pavedRoadService.getPersistedAssetsByLinkIds(PavedRoad.typeId, Seq(link.linkId))
+          duplicate.isEmpty match {
+            case false =>
+              pavedRoadService.update(Seq(duplicate.head.id), pavementClassValue, AwsUser)
+            case true =>
+              val newPavementClassAsset = NewLinearAsset(link.linkId, 0, link.length, pavementClassValue, properties("sideCode").asInstanceOf[BigInt].intValue(), vvhClient.roadLinkData.createVVHTimeStamp(), None)
+              pavedRoadService.create(Seq(newPavementClassAsset), PavedRoad.typeId, AwsUser)
+          }
+      }
+    }
+  }
+
+  def validateAndInsertDataset(dataset: Dataset): Int = {
+    var featuresWithoutIds = 0
+
+    var datasetStatus = 0
+    val assets = dataset.geoJson("features").asInstanceOf[List[Map[String, Any]]]
+    val roadlinks = dataset.roadlinks
+
+    if (assets.length != dataset.roadlinks.length) {
+      datasetStatus = 1
+    }
+
+    awsDao.insertDataset(dataset.datasetId, write(dataset.geoJson), write(dataset.roadlinks), datasetStatus)
+    if (datasetStatus != 1) {
+      var roadlinksCount = 0
+
+      assets.foreach(feature => {
+        val featureRoadlinks = roadlinks(roadlinksCount)
+        roadlinksCount += 1
+
+        val properties = feature.getOrElse("properties", "").asInstanceOf[Map[String, Any]]
+        properties.get("id") match {
+          case Some(featureId) =>
+            try {
+              awsDao.insertFeature(featureId.asInstanceOf[BigInt].longValue(), dataset.datasetId, 0)
+
+              linkIdValidation(featureRoadlinks.map(number => number.longValue()).toSet)
+
+              val assetTypeGeometry = feature("geometry").asInstanceOf[Map[String, Any]]("type")
+              assetTypeGeometry match {
+                case "LineString" =>
+                  properties("type").asInstanceOf[String] match {
+                    case "Roadlink" => validateRoadlink(dataset.datasetId, featureId.asInstanceOf[BigInt].longValue(), properties)
+                    case _ =>
+                      awsDao.updateFeatureStatus(featureId.asInstanceOf[BigInt].longValue(), "1")
```

Comment:
You know the type of error so create a new code for this error instead of using the generic one